base/set.jl

```diff
@@ -117,6 +117,9 @@ copymutable(s::Set{T}) where {T} = Set{T}(s)
 # Set is the default mutable fall-back
 copymutable(s::AbstractSet{T}) where {T} = Set{T}(s)
 
+function _maybe_sizehint!(s::Set, size::Int)
+    size*3 > length(s.dict.keys) * 2 && sizehint!(s, size)
+end
```

Comment:
Can we alter `sizehint!` directly to include this check, rather that changing it here? It looks like currently it rounds it up to the next power of 2 instead.

Reply:
Yes, it rounds to `2^n` here https://github.com/JuliaLang/julia/blob/6f983aaa26db234b4feadcc6bc3d1f2e6166a4b8/base/dict.jl#L232

I'd prefer to implement all the resizing logic at a single place in `sizehint!`. If we do not want to expose new argument to public api, we can implement it in an internal funcition, like
```julia
sizehint!(d::Dict{T}, newsz) where T  = _sizehint!(d, newsz) 

function _sizehint!(d::Dict{T}, newsz; shrink = true) where T
    oldsz = length(d.slots)
    # limit new element count to max_values of the key type
    newsz = min(max(newsz, length(d)), max_values(T)::Int)
    # need at least 1.5n space to hold n elements
    newsz = _tablesz(cld(3 * newsz, 2))
    noupdate = shrink : newsz == oldsz ? newsz <= oldsz
    return noupdate ? d : rehash!(d, newsz)
end
```